[WIP] Add end-to-end test for RNN Module

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29543 [WIP] Add end-to-end test for RNN Module**
* #29781 Always include autograd context id in rpc/remote requests
* #29634 Allow rpc.remote to create RRef on self

The test put encoder and decoder on a remote worker, and put the
LSTM module locally. The forward pass 1st looks up the embedding
remotely, then fetchs the emb result and runs it through the local
LSTM module, and finally sends the output to the remote decoder.
The backward pass should automatically traverse through all
involved parties. The optimizer takes a list of param RRefs, and
reaches each owner to update the params.

Differential Revision: [D18482428](https://our.internmc.facebook.com/intern/diff/D18482428)

mrshenli committed:
[WIP] Add end-to-end test for RNN Module

The test put encoder and decoder on a remote worker, and put the
LSTM module locally. The forward pass 1st looks up the embedding
remotely, then fetchs the emb result and runs it through the local
LSTM module, and finally sends the output to the remote decoder.
The backward pass should automatically traverse through all
involved parties. The optimizer takes a list of param RRefs, and
reaches each owner to update the params.

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -162,7 +162,10 @@ std::shared_ptr<FutureMessage> pyRpcPythonUdf(
+      agent,
+      dst,
+      std::move(*pythonCall).toMessage(),
+      true /*forceGradRecording*/);
```
Adding `forceGradRecording` because UDFs might not carry any requires_grad tensor, but does requires grad on the return value.

pritamdamania87 replied:
IIUC, this fixes https://github.com/pytorch/pytorch/issues/28819. If so, can we have a simple unit test for the problem this fixes?

mrshenli replied:
Yes, actually let me split this PR into two, with the first one focusing on fixing #28819 with unit test.

mrshenli committed:
Update on "[WIP] Add end-to-end test for RNN Module"


The test put encoder and decoder on a remote worker, and put the
LSTM module locally. The forward pass 1st looks up the embedding
remotely, then fetchs the emb result and runs it through the local
LSTM module, and finally sends the output to the remote decoder.
The backward pass should automatically traverse through all
involved parties. The optimizer takes a list of param RRefs, and
reaches each owner to update the params.

[ghstack-poisoned]

mrshenli commented on test/dist_model_parallel_test.py:
```diff
@@ -0,0 +1,138 @@
+            inp = torch.LongTensor(batch, nindices) % ntoken
+            output, hidden = rnn(inp, hidden)
+            dist_autograd.backward([output.sum()])
+            #opt.step()
```
cc @soumith 

This would be what we expect users to write to train an RNN module. Will remove comments on `DistributedOptimizer` once #29304 and #28948 are in.

mrshenli committed:
Update on "[WIP] Add end-to-end test for RNN Module"


The test put encoder and decoder on a remote worker, and put the
LSTM module locally. The forward pass 1st looks up the embedding
remotely, then fetchs the emb result and runs it through the local
LSTM module, and finally sends the output to the remote decoder.
The backward pass should automatically traverse through all
involved parties. The optimizer takes a list of param RRefs, and
reaches each owner to update the params.

[ghstack-poisoned]

rohan-varma commented on test/dist_model_parallel_test.py:
```diff
@@ -0,0 +1,138 @@
+        args=[RemoteModule.remote_parameters, remote_module_rref]
+    )
+
+class Encoder(RemoteModule):
```
nit, should this be called something like `EmbeddingLayer` or `RemoteEmbeddingLayer`? Usually in seq2seq models encoder refers to the RNN that creates the encoded representation, which in this case would be the LSTM.

aazzolini replied:
RemoteModule here is a misnomer, this is a perfectly well defined local module actually. It just happens to be used  as a remote module somewhere. Nothing in the implementation of this module implies remote.

pritamdamania87 commented on test/dist_model_parallel_test.py:
```diff
@@ -0,0 +1,138 @@
+# from torch.distributed.optim import DistributedOptimizer
+# from torch.distributed.rpc import RRef
+
+from dist_utils import INIT_METHOD_TEMPLATE, dist_init, TEST_CONFIG
```
Looks like we have a bunch of lint failures.

pritamdamania87 commented on test/dist_model_parallel_test.py:
```diff
@@ -0,0 +1,138 @@
+import torch
```
Wasn't the plan to have this in the [examples](https://github.com/pytorch/examples) repository? Why is this a unit test instead?

mrshenli replied:
We need RPC module tests as well. I would first like to check with @soumith on whether this is a reasonable example, and whether there are any APIs that we need to revise. And also waiting for the dist optimizer to be landed (it's landing now).

pritamdamania87 commented on test/test_dist_model_parallel_spawn.py:
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env python3
+from __future__ import absolute_import, division, print_function, unicode_literals
```
Why do we have two separate files if we only have a `spawn` mode?

mrshenli replied:
Let me merge this into one file.

pritamdamania87 commented on torch/csrc/distributed/autograd/utils.cpp:
```diff
@@ -47,10 +47,7 @@ DistAutogradContext* addRecvRpcBackward(
-    TORCH_INTERNAL_ASSERT(
-        torch::autograd::compute_requires_grad(tensors),
-        "Received tensors do not require grad, addRecvRpcBackward should not be called");
+  if (!tensors.empty() && torch::autograd::compute_requires_grad(tensors)) {
```
Were the changes here and in python_functions.cpp bugs that were discovered by the unit test in this PR?

mrshenli replied:
Yes, I hit errors when running unit test in this PR, but it's kind of expected as we already have #28819 to track this.

aazzolini commented on test/dist_model_parallel_test.py:
```diff
@@ -0,0 +1,138 @@
+        kwargs=kwargs
+    )
+
+class RemoteModule(nn.Module):
```
What is the reason for RemoteModule in here? I believe RemoteModule only makes sense as a wrapper module that wraps regular modules. The way it's exposed here you'll still have to implement a special module that inherits from RemoteModule, so there's not much utility for this class.

If the only functionality is exposing remote_parameters(), then simply make it a free function, e.g.:

def get_module_param_rrefs(nn_module: NNModule):
    return [RRef(param) for param in self.parameters()]

If the goal is to discover remote modules, you could have a RRef wrapper on the client side instead (not on the remote side).

aazzolini commented on test/dist_model_parallel_test.py:
```diff
@@ -0,0 +1,138 @@
+        return self.drop(self.encoder(input))
+
+
+class RNN(RemoteModule):
```
Same here

aazzolini reviewed (commented):
I believe if we do use  RemoteModule, it should be a wrapper that lives on the "client" side, not on the "server" side. The way the example is presented here, I'd rather just not introduce the concept of RemoteModule and just call into nn.Module directly.

mrshenli committed:
Update on "[WIP] Add end-to-end test for RNN Module"


The test put encoder and decoder on a remote worker, and put the
LSTM module locally. The forward pass 1st looks up the embedding
remotely, then fetchs the emb result and runs it through the local
LSTM module, and finally sends the output to the remote decoder.
The backward pass should automatically traverse through all
involved parties. The optimizer takes a list of param RRefs, and
reaches each owner to update the params.

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Add end-to-end test for RNN Module"


The test put encoder and decoder on a remote worker, and put the
LSTM module locally. The forward pass 1st looks up the embedding
remotely, then fetchs the emb result and runs it through the local
LSTM module, and finally sends the output to the remote decoder.
The backward pass should automatically traverse through all
involved parties. The optimizer takes a list of param RRefs, and
reaches each owner to update the params.

Differential Revision: [D18482428](https://our.internmc.facebook.com/intern/diff/D18482428)

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Add end-to-end test for RNN Module"


The test put encoder and decoder on a remote worker, and put the
LSTM module locally. The forward pass 1st looks up the embedding
remotely, then fetchs the emb result and runs it through the local
LSTM module, and finally sends the output to the remote decoder.
The backward pass should automatically traverse through all
involved parties. The optimizer takes a list of param RRefs, and
reaches each owner to update the params.

Differential Revision: [D18482428](https://our.internmc.facebook.com/intern/diff/D18482428)

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Add end-to-end test for RNN Module"


The test put encoder and decoder on a remote worker, and put the
LSTM module locally. The forward pass 1st looks up the embedding
remotely, then fetchs the emb result and runs it through the local
LSTM module, and finally sends the output to the remote decoder.
The backward pass should automatically traverse through all
involved parties. The optimizer takes a list of param RRefs, and
reaches each owner to update the params.

Differential Revision: [D18482428](https://our.internmc.facebook.com/intern/diff/D18482428)

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Add end-to-end test for RNN Module"


The test put encoder and decoder on a remote worker, and put the
LSTM module locally. The forward pass 1st looks up the embedding
remotely, then fetchs the emb result and runs it through the local
LSTM module, and finally sends the output to the remote decoder.
The backward pass should automatically traverse through all
involved parties. The optimizer takes a list of param RRefs, and
reaches each owner to update the params.

Differential Revision: [D18482428](https://our.internmc.facebook.com/intern/diff/D18482428)

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Add end-to-end test for RNN Module"


The test put encoder and decoder on a remote worker, and put the
LSTM module locally. The forward pass 1st looks up the embedding
remotely, then fetchs the emb result and runs it through the local
LSTM module, and finally sends the output to the remote decoder.
The backward pass should automatically traverse through all
involved parties. The optimizer takes a list of param RRefs, and
reaches each owner to update the params.

Differential Revision: [D18482428](https://our.internmc.facebook.com/intern/diff/D18482428)

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Add end-to-end test for RNN Module"


The test put encoder and decoder on a remote worker, and put the
LSTM module locally. The forward pass 1st looks up the embedding
remotely, then fetchs the emb result and runs it through the local
LSTM module, and finally sends the output to the remote decoder.
The backward pass should automatically traverse through all
involved parties. The optimizer takes a list of param RRefs, and
reaches each owner to update the params.

Differential Revision: [D18482428](https://our.internmc.facebook.com/intern/diff/D18482428)

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Add end-to-end test for RNN Module"


The test put encoder and decoder on a remote worker, and put the
LSTM module locally. The forward pass 1st looks up the embedding
remotely, then fetchs the emb result and runs it through the local
LSTM module, and finally sends the output to the remote decoder.
The backward pass should automatically traverse through all
involved parties. The optimizer takes a list of param RRefs, and
reaches each owner to update the params.

Differential Revision: [D18482428](https://our.internmc.facebook.com/intern/diff/D18482428)

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Add end-to-end test for RNN Module"


The test put encoder and decoder on a remote worker, and put the
LSTM module locally. The forward pass 1st looks up the embedding
remotely, then fetchs the emb result and runs it through the local
LSTM module, and finally sends the output to the remote decoder.
The backward pass should automatically traverse through all
involved parties. The optimizer takes a list of param RRefs, and
reaches each owner to update the params.

Differential Revision: [D18482428](https://our.internmc.facebook.com/intern/diff/D18482428)

[ghstack-poisoned]

rohan-varma commented on test/test_dist_model_parallel_spawn.py:
```diff
@@ -0,0 +1,18 @@
+import unittest
+
+@unittest.skipIf(TEST_WITH_ASAN, "Skip ASAN as torch + multiprocessing spawn have known issues")
+class DistModelParallelSpawn(MultiProcessTestCase, DistModelParallelTest):
```
curious, is there a reason we don't have a fork mode test for this?

rohan-varma commented:
Are we still planning to merge this eventually? Would be super useful to have an end to end test to sanity check changes we make in the RPC layer.
